File: CSharpCodeBase/utils/bhlib/statefulsequence.lua
local Class = require 'utils.hump.class'

local StatefulSequence = Class({init = function(self, behaviours)
      self._behaviours = behaviours
      self.returnCode = B_FAILURE
      self.lastBehaviour = 1
end})

function StatefulSequence:Behave(creature)
  while self.lastBehaviour <= #self._behaviours do
    code = self._behaviours[self.lastBehaviour]:Behave(creature)
    if code == B_FAILURE then
      self.lastBehaviour = 1
      self.returnCode = code
      return self.returnCode
    elseif code == B_SUCCESS then
      -- continue
    elseif code == B_RUNNING then
      self.returnCode = code
      return self.returnCode
    end
    self.lastBehaviour = self.lastBehaviour + 1
  end
  
  self.lastBehaviour = 1
  self.returnCode = B_SUCCESS
  return self.returnCode
end

return StatefulSequence

File: CSharpCodeBase/utils/bhlib/statefulsequence.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class statefulsequence {
 public void init(self, behaviours);
       this._behaviours = behaviours;
       this.returnCode = B_FAILURE;
       this.lastBehaviour = 1;
 }})
 public void Behave(creature){
   while(this.lastBehaviour <= #this._behaviours ){
     code = this._behaviours[this.lastBehaviour]:Behave(creature);
     if(code == B_FAILURE  ){ 
       this.lastBehaviour = 1;
       this.returnCode = code;
       return this.returnCode;
     }else{if code == B_SUCCESS  ){ 
       // continue
     }else{if code == B_RUNNING  ){ 
       this.returnCode = code;
       return this.returnCode;
     }
     this.lastBehaviour = this.lastBehaviour + 1;
   }
   
   this.lastBehaviour = 1;
   this.returnCode = B_SUCCESS;
   return this.returnCode;
 }
 }}